lobster/related.lobster

```diff
@@ -0,0 +1,88 @@
+
+import std
+import dictionary
+import json
+
+let TOPN = 5
+
+struct Post:
+    _id: string
+    title: string
+    tags: [string]
+
+struct RelatedPost:
+    _id: string
+    tags: [string]
+    related: [Post]
+
+def main():
+    let json_content = read_file("../posts.json")
+
+    assert(json_content != nil)
+
+    let posts = JSON.parse<[Post]>(json_content)
+
+    let start = seconds_elapsed()
+
+    let tag_map = dictionary<string, [int]> ( 100 )
+
+    for (posts) post,i:
+        for (post.tags) tag:
+            let indexes = tag_map.get(tag)
+
+            if not indexes:
+                tag_map.set(tag, [i])
+            else:
+                tag_map.set(tag, push(indexes, i))
+
+    var all_related_posts = []::RelatedPost
+    let posts_count = posts.length
+
+    vector_capacity(all_related_posts, posts_count)
+
+    for(posts) post,i:
+
+        let tagged_post_count = map(posts_count) : 0
+
+        for (post.tags) tag:
+            for (assert(tag_map.get(tag))) index:
+                tagged_post_count[index] += 1
+        
+        tagged_post_count[i] = 0
+
+        let topn = map(TOPN*2) : 0
+        var min_tags = 0
+
+        for(tagged_post_count) count, idx:
+            guard count > min_tags
+
+            var upper_bound = (TOPN-2) * 2
+
+            while upper_bound >= 0 and count > topn[upper_bound]:
+                topn[upper_bound+2] = topn[upper_bound]
```

Comment:
This seems to be the most inner of loops? Have you checked for other languages what percentage of time is spent in this loop? Because you risk your benchmark is actually mostly testing the language's ability to do array indexing..

Reply:
I didn't do any profiling outside of go, this approach was slower in go so a different algo was used. Rust's binary heap was faster that both.